<a href="https://colab.research.google.com/github/LKXoro/BEDU-Rstudio-postworks/blob/main/M3_Postwork4/Postwork_04_python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Postwork 04: Análisis Exploratorio de nuestro Dataset

En el [postwork anterior](https://github.com/LKXoro/BEDU-Rstudio-postworks/blob/main/M3_Postwork3/Postwork_03_python.ipynb) exploramos las posibles respuestas que nos brinda PandaScore. Con ellas obtuvimos 5 Datasets llenos de información tan sólo para League of Legends. Ahora vamos a inspeccionar un poco los campos que tenemos y el tipo de información dentro de ellos, comencemos realizando las peticiones a la API:

In [2]:
import requests 

juego = 'lol' 
información = 'tournaments' 
información2 = 'leagues'
información3 = 'series' 
información4 = 'matches'
información5 = 'teams'
registros = 100  
token =  'V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns'

url = f"https://api.pandascore.co/{juego}/{información}?page[size]={registros}&token=V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns"
url2 = f"https://api.pandascore.co/{juego}/{información2}?page[size]={registros}&token=V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns"
url3 = f"https://api.pandascore.co/{juego}/{información3}?page[size]={registros}&token=V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns"
url4 = f"https://api.pandascore.co/{juego}/{información4}?page[size]={registros}&token=V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns"
url5 = f"https://api.pandascore.co/{juego}/{información5}?page[size]={registros}&token=V3U1fnnWozJi3bbhZfrxdiAmN4-XWnX_4g6i_6iauwwh7w-V8Ns"

response = requests.get(url) 
response2 = requests.get(url2)
response3 = requests.get(url3) 
response4 = requests.get(url4) 
response5 = requests.get(url5) 

print(response.status_code)
print(response2.status_code)
print(response3.status_code)
print(response4.status_code)
print(response5.status_code)


200
200
200
200
200


¡Genial! Todas las peticiones se realizaron correctamente. Vamos a explorar un poco la cantidad de datos que se nos devolvió:

In [3]:
print(f'### La petición de {información} nos devolvió:\n')
print(response.headers['X-Total']+' registros totales.\n') 
print(f'### La petición de {información2} nos devolvió:\n')
print(response2.headers['X-Total']+' registros totales.\n') 
print(f'### La petición de {información3} nos devolvió:\n')
print(response3.headers['X-Total']+' registros totales.\n') 
print(f'### La petición de {información4} nos devolvió:\n')
print(response4.headers['X-Total']+' registros totales.\n') 
print(f'### La petición de {información5} nos devolvió:\n')
print(response5.headers['X-Total']+' registros totales.\n')   

### La petición de tournaments nos devolvió:

824 registros totales.

### La petición de leagues nos devolvió:

67 registros totales.

### La petición de series nos devolvió:

298 registros totales.

### La petición de matches nos devolvió:

18694 registros totales.

### La petición de teams nos devolvió:

973 registros totales.



Observamos que los datos relacions con los enfrentamientos (matches) cuentan con una extensa cantidad de registros, esto es lógico, ya que dentro de cada torneo ha habido diversos enfrentamientos (¡Y han habido muchos torneos!). Exploremos mejor los campos de cada respuesta, primero creamos los df:

In [8]:
import pandas as pd

data1 = response.json()
data2 = response2.json()
data3 = response3.json()
data4 = response4.json()
data5 = response5.json()

df_torneos = pd.DataFrame.from_dict(data1)
df_ligas= pd.DataFrame.from_dict(data2)
df_series = pd.DataFrame.from_dict(data3)
df_enfrentamientos = pd.DataFrame.from_dict(data4)
df_equipos = pd.DataFrame.from_dict(data5)

Si seguimos la jerarquía de organización de los eventos de eSports, el primer elemento que debemos inspeccionar son las ligas:

In [10]:
df_ligas.head()

id  ...                                          videogame
0  4539  ...  {'current_version': '11.5.1', 'id': 1, 'name':...
1  4533  ...  {'current_version': '11.5.1', 'id': 1, 'name':...
2  4483  ...  {'current_version': '11.5.1', 'id': 1, 'name':...
3  4450  ...  {'current_version': '11.5.1', 'id': 1, 'name':...
4  4433  ...  {'current_version': '11.5.1', 'id': 1, 'name':...

[5 rows x 8 columns]

Nos encontramos ante un df con 8 campos, dos de ellos son url que nos llevan a la página oficial (en caso de tenerla) y el logo oficial de la serie. También tenemos un campo para la versión del juego con la que se llevo a cabo la liga y las distintas series que se han llevado a cabo, el _slug_ (¿eslogan?) de la liga y el nombre de la liga. 

La cantidad de datos sobre ligas, al ser los eventos de mayor nivel, es la de menor cantidad. El número de campos es algo limitado y no todos son de útilidad para análisis, para un dashboard los url e imágenes serían ideales.

Veamos que información tenemos respecto a las series:

In [18]:
df_series.head()

begin_at description  ... winner_type  year
0  2021-02-23T07:00:00Z        None  ...        None  2021
1  2021-01-25T18:00:00Z        None  ...        None  2021
2  2021-02-13T13:00:00Z        None  ...        None  2021
3  2021-02-19T08:00:00Z        None  ...        None  2021
4  2021-01-19T19:00:00Z        None  ...        None  2021

[5 rows x 18 columns]

En este Dataframe encontramos más campos como la fecha de inicio y fin de la serie, un campo de descripción que parece estar lleno de valores _None_ al igual que el nombre del juego, el nombre de la serie y slug, la temporada y nivel y se la serie,encontramos también el id de la liga correspondiente y el id del ganador (si no se ha definido un ganador se tiene un valor NaN), el tipo de ganador (LoL es un juego de equipos así que en general el tipo de ganador es _Team_ y también tenemos el año al que corresponde la serie.

Dentro de estos campos también tenemos el que corresponde al de los torneos, veamos que tenemos respecto a ellos:

In [17]:
df_torneos.head()

begin_at                end_at  ...  winner_id winner_type
0  2021-03-09T17:00:00Z                  None  ...        NaN        Team
1  2021-03-09T16:30:00Z                  None  ...        NaN        Team
2  2021-03-09T18:00:00Z  2021-03-23T23:00:00Z  ...        NaN        Team
3  2021-03-29T08:00:00Z  2021-04-09T14:00:00Z  ...        NaN        Team
4  2021-03-27T16:00:00Z  2021-04-18T21:00:00Z  ...        NaN        Team

[5 rows x 17 columns]

Encontramos las fechas de inicio y fin de cada torneo, su id de torneo junto a los id de la liga y serie a la que corresponde. Fecha de última actualización, soporte de stream en vivo, nombre del torneo, _slug_ y _prizepool_ que parece no tener un valor diferente a _None_, tenemos el winner id y winner type (en caso de haber un ganador) y dos campos dedicados a los enfrentamientos y equipos. 

En los enfrentamientos encontramos aún más información:

In [24]:
df_enfrentamientos.head()

name  ...              begin_at
0            Round 3: TBD vs TBD  ...  2021-03-11T17:00:00Z
1   Round 2 Match 2: MOUZ vs BIG  ...  2021-03-10T20:00:00Z
2    Round 2 Match 1: TBD vs TBD  ...  2021-03-10T17:00:00Z
3  Round 1 Match 2: SKP vs S04.E  ...  2021-03-09T20:00:00Z
4     Round 1 Match 1: GL vs USE  ...  2021-03-09T17:00:00Z

[5 rows x 33 columns]

Parece que este Dataframe contiene información que nos puede ayudar a analizar el _performance_ de los equipos y además tenemos datos sobre las partidas y sus streams. Encontramos las fechas de inicio y fin de los enfrentamientos, los id de ligas y torneos, resultados, la versión del juego con la que se jugó, tipos de enfrentamiento, número de partidas por enfrentamiento, si hubo abandonó o cambios de calendarización, etc.

Hasta el momento los datos respecto a los enfrentamientos parecen aportar más detalles. Finalmente, echemos un ojo a los equipos:

In [25]:
df_equipos.head()

acronym  ...                        slug
0     PCE  ...                       peace
1     Z10  ...                  z10-eshark
2     X25  ...                 x25-esports
3    None  ...                   ctrl-play
4     IPN  ...  impunity-league-of-legends

[5 rows x 9 columns]

Al igual que el Dataframe de las ligas, los datos de los equipos parecen estar más limitados a sus nombres, logotipos promocionales, páginas web y algún que otro dato de sus integrantes. Después de está inspección preliminar parece claro que la información más _jugosa_ la encontramos directamente en el calor de la batalla, es decir, los **enfrentamientos**.

Inspeccionemos sus tipos de datos:

In [30]:
df_enfrentamientos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 33 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   name                   100 non-null    object 
 1   official_stream_url    93 non-null     object 
 2   end_at                 31 non-null     object 
 3   draw                   100 non-null    bool   
 4   league                 100 non-null    object 
 5   league_id              100 non-null    int64  
 6   detailed_stats         100 non-null    bool   
 7   live_embed_url         93 non-null     object 
 8   tournament_id          100 non-null    int64  
 9   results                100 non-null    object 
 10  videogame_version      31 non-null     object 
 11  streams                100 non-null    object 
 12  modified_at            100 non-null    object 
 13  games                  100 non-null    object 
 14  original_scheduled_at  100 non-null    object 
 15  match_t

Recordemos que este Dataframe es solo una pequeña parte de todos los datos que nos puede entregar la API, 100 registros de 18694 en realidad, pero podemos sacar algunas conclusiones preliminares.

Es necesario transformar los campos de fechas al tipo de dato correspondiente para su manejo, tal vez sea necesario eliminar algunas columnas que sea redundantes o que no aporten información útil, como _game advantage_ que parece estar lleno de valores _None_. Puede que sea conveniente transformar algunos datos a tipo _string_ para obtener más información.

Una de las cosas que pueden resultar problemáticas son los enfrentamientos que estan _agendados_ pero no se han llevado a cabo, si bien es bueno tenerlos en cuenta, muchos de los campos terminan teniendo valores nulos a causa de estos, por lo que resulta conveniente pensar solo en aquellos enfrentamientos que ya se han llevado a cabo.

Hay muchas posibilidades, pero por el momento, parece que el dataframe que mejor podemos explotar para responder a las preguntas que nos hicimos es el correspondiente a los enfrentamientos.